scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncDoFn.java

```diff
@@ -71,24 +71,20 @@ public void processElement(ProcessContext c, BoundedWindow window) {
     flush(c);
 
     final UUID uuid = UUID.randomUUID();
-    FutureT future =
-        addCallback(
-            processElement(c.element()),
-            r -> {
-              results.add(new Result(r, c.timestamp(), window));
-              futures.remove(uuid);
-              return null;
-            },
-            t -> {
-              errors.add(t);
-              futures.remove(uuid);
-              return null;
-            });
-    // This `put` may happen after `remove` in the callbacks but it's OK since either the result
-    // or the error would've already been pushed to the corresponding queues and we are not losing
-    // data. `waitForFutures` in `finishBundle` blocks until all pending futures, including ones
-    // that may have already completed, and `startBundle` clears everything.
-    futures.put(uuid, future);
+    futures.computeIfAbsent(
+      uuid,
+      key -> addCallback(
+        processElement(c.element()),
+        r -> {
+          results.add(new Result(r, key, c.timestamp(), window));
+          return null;
+        },
+        t -> {
+          errors.add(t);
```

Comment:
Do we need to `futures.remove(key)` here since if it fails, it'll never be added to `results` and leaked inside the map?

Reply:
Although I guess since errors will cause the DoFn to fail and GCed?

Reply:
Correct, this will cause the DoFn to fail, so it'll be automatically GCed I believe